Fix some memory bugs in `runtime_events_consumer.c`

eutro opened the pull request:
This PR fixes a few memory bugs surrounding `runtime_events_loc` in `runtime_events_consumer.c`, as mentioned in #13089, specifically in [`caml_runtime_events_create_cursor`](https://github.com/ocaml/ocaml/blob/4c6a3849022ba19c23fb1860095f65eb09da157c/otherlibs/runtime_events/runtime_events_consumer.c#L92).

These bugs are:

- The code path for `Runtime_events.create_cursor None` (the `pid < 0` branch in the C code) allocates `runtime_events_loc` but reassigns it, making the old pointer unreachable without deallocating it:

https://github.com/ocaml/ocaml/blob/4c6a3849022ba19c23fb1860095f65eb09da157c/otherlibs/runtime_events/runtime_events_consumer.c#L108-L124

- The function frees `runtime_events_loc` after it is introduced if and only if the function returns with an error.[^1] This is incorrect, both because:
  - in the `pid < 0` code path, it shouldn't be freed at all, since [`caml_runtime_events_current_location()` does not return a new string](https://github.com/ocaml/ocaml/blob/4c6a3849022ba19c23fb1860095f65eb09da157c/runtime/runtime_events.c#L222)
  - in the `pid >= 0` code path, it should always be freed, since `runtime_events_loc` does not escape the function

This PR:

- Fixes these by moving `runtime_events_loc`'s allocation into the relevant branch, rewriting the function to be single-exit after cursor allocation succeeds, and by having both frees state their explicit guards:

https://github.com/ocaml/ocaml/blob/1c337f6d56604663f8b2eb839fe14abfb8230856/otherlibs/runtime_events/runtime_events_consumer.c#L237-L252

- Deduplicates the two identical `path`-freeing if-blocks in the `caml_ml_runtime_events_create_cursor` wrapper which calls it, for clarity. This does not change the behaviour of the program.
- Adds a test which tries to reach these cases, though only the double-free (and Windows on the commit before [#13089](/ocaml/ocaml/pull/13089)) actually causes a failure of the test.
  - This `chmod 000`s the ring buffer file in order to cause subsequent read attempts to fail.
  - It uses a slight workaround to find the correct PID and `.events` file, so it works on Windows (rather than skipping), because I wanted to actually test [#13089](/ocaml/ocaml/pull/13089).
  - The testsuite does fail on this commit.

[^1]: This is the correct behaviour for `cursor`, since the caller takes ownership of the returned cursor iff the function returns successfully

eutro committed:
Clarify and fix `caml_stat` memory management in `runtime_events_consumer.c`

eutro committed:
Add tests/lib-runtime-events/test_create_cursor_failures.ml

Test cases for some recent bugs in lib-runtime-events

eutro commented:
It looks like, perhaps unsurprisingly, not all platforms work with the new test (though MSVC just had a quite disappointing network error), so I intend to remove it later, since I can't think of a better way to force failure.

dustanddreams commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -100,28 +100,31 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
 
   struct caml_runtime_events_cursor *cursor =
       caml_stat_alloc_noexc(sizeof(struct caml_runtime_events_cursor));
+  int should_free_events_loc;
```
I'd rather see that variable initialized to zero at the declaration point, so that future changes in this routine do not risk using an uninitialized value.

tmcgilchrist replied:
Following in that style should `int ret;` get initialised to something since it is being returned at the end of the function?

dustanddreams replied:
That wouldn't hurt either.

eutro replied:
I made `ret` uninitialised specifically so the compiler would emit a warning if it isn't assigned before jumping; and `should_free_events_loc` uninitialised because it's to be initialised where `runtime_events_loc` is.

A change that would accidentally use these with the default initialisation is already wrong, so my reasoning is that it's better to have an obviously invalid value that the compiler would yell at us about.

dustanddreams replied:
This is relying upon the C compiler being always correct in figuring out whether a variable is initialized or not. It turns out they aren't always, with both false positives and false negatives. I wouldn't put too much trust in the compiler.

gasche replied:
Drive-by comment: do we need a boolean to tell us to free `runtime_events_loc`, or could we just free exactly when the variable is not `NULL`?

eutro replied:
> Drive-by comment: do we need a boolean to tell us to free runtime_events_loc, or could we just free exactly when the variable is not NULL?

It's not the `NULL` case we're worried about, but the case where it's initialised to `caml_runtime_events_current_location()`, which isn't a copy.

gasche replied:
I see. I looked out of curiosity, wondering if just making a copy in that case would make things nicer. My quick skim suggests that the function is sprawling and somewhat of a mess, and I think that this is going to bite us again in the future. So I would encourage you to refactor it as you see fit -- for example maybe the system-dependent bits could be factored out into their own helper functions -- to make the code nicer, and not just safer. Your current change is okay but it does not make the code nicer to read.

eutro replied:
I've refactored `caml_runtime_events_create_cursor` and made `runtime_events_loc` be a new string every time to make reasoning easier.

tmcgilchrist commented on testsuite/tests/lib-runtime-events/test_create_cursor_failures.ml:
```diff
@@ -0,0 +1,57 @@
+  Runtime_events.pause ()
+
+(* workaround for finding the events file even on Windows, where
+   [Unix.getpid] doesn't match the one used to open the file *)
```
Can you explain what the difference is on Windows? The header file  says:
> [pid] is the process id (or equivalent) of the startup OCaml process.

https://github.com/ocaml/ocaml/blob/trunk/otherlibs/runtime_events/caml/runtime_events_consumer.h#L31

If there is a known difference it would be useful to state that in the header file.

eutro replied:
The issue is not with Windows or `Runtime_events`, but with the way the `Unix` library handles PIDs on Windows (as Windows `HANDLE`s casted to `int`). `Unix.getpid()` doesn't actually return the PID of the process (see #4034), whereas the ring buffer file does use the actual PID. I mention possible fixes in [this footnote](https://github.com/ocaml/ocaml/pull/13089#user-content-fn-2-dc20ce2080ddb24747cbd553a6236931) on #13089.

eutro committed:
Refactor `caml_runtime_events_create_cursor`

- Introduce `format_runtime_events_loc`, allocating a new string each time

- Introduce `cursor_map_ring_file`, which now also closes `ring_fd`, and closes
handles on Windows on failure

- Use `memset` to zero-initialise cursor callbacks

eutro committed:
Swap `ring_handle` and `ring_file_handle` frees in Windows cursor mapping

eutro commented:
I've spotted and fixed another bug, notably `Runtime_events.create_cursor None` never closing the file descriptor on Unix.

On trunk this crashes after opening too many file descriptors, causing `create_cursor None` to fail and trigger the existing double-free bug:

```ocaml
let () =
  Runtime_events.start ();
  try
    for _ = 1 to 1024 (* or whatever your [ulimit -n] is *) do
      Runtime_events.(create_cursor None |> free_cursor)
    done
  with _ ->
    Runtime_events.(create_cursor None |> free_cursor)
```

I also perform cleanup on the Windows handles if the mapping fails.

NickBarnes commented:
I'll review this.

MisterDA commented:
Should the ring file be also marked non-inheritable on Windows and close-on-exec on Unix?

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -88,57 +88,47 @@ struct caml_runtime_events_cursor {
-  struct caml_runtime_events_cursor *cursor =
-      caml_stat_alloc_noexc(sizeof(struct caml_runtime_events_cursor));
+/** Return a new string with the path of the ring file */
+static runtime_events_error format_runtime_events_loc(
```
Note that (because C is antediluvian) enumeration constants such as `E_PATH_FAILURE` are of type `int`, so here you are forcing an implicit conversion into the enum type, and then at the call site there's an implicit conversion back to `int`. So maybe this return type should be `int`? For this reason, I never name enum types (so I never have variables or slots of enum types).

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
-  }
+ failed2:
+  CloseHandle(cursor->ring_handle);
+ failed1:
```
I suggest using fail labels which reflect the failure, e.g. `fail_file_mapping` and `fail_map_view` here.

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -149,10 +139,7 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
   if (cursor->ring_file_handle == INVALID_HANDLE_VALUE) {
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
     return E_OPEN_FAILURE;
```
For consistency, please set `ret` and then `goto fail_create_file` here (putting that label right before the `return ret`).

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -88,57 +88,47 @@ struct caml_runtime_events_cursor {
+                     char_os *runtime_events_loc) {
+  int ret = 0;
 #ifdef _WIN32
   cursor->ring_file_handle = CreateFile(
```
Strong preference for not setting any fields in `cursor` until we know we're succeeding. Use local variables and then copy them in right before `return E_SUCCESS`.

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
+ failed1:
+  CloseHandle(cursor->ring_file_handle);
+  return ret;
+#else
```
The stylistic distinction between the Windows and non-Windows sides of this function is quite jarring.

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
-    caml_stat_free(runtime_events_loc);
-    return E_OPEN_FAILURE;
+    ret = E_OPEN_FAILURE;
+    goto failed;
```
`fail_fstat`

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
+  return ret;
+#else
+  int ring_fd = open(runtime_events_loc, O_RDONLY, 0);
+  if(ring_fd == -1) return E_OPEN_FAILURE;
```
`ret = E_OPEN_FAILURE; goto fail_open;`

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
+    goto failed;
   }
-
   cursor->ring_file_size_bytes = tmp_stat.st_size;
```
Local variables please (see comment on Windows side).

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
+    mmap(NULL, cursor->ring_file_size_bytes, PROT_READ, MAP_SHARED, ring_fd, 0);
+  if(cursor->metadata == MAP_FAILED) {
+    ret = E_MAP_FAILURE;
+    goto failed;
```
`fail_map`

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
   }
+
+  ret = E_SUCCESS;
+  /* fallthrough */
```
There is no fall-through.

NickBarnes replied:
I was mistaken!

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
   *cursor_res = cursor;
+  ret = E_SUCCESS;
+  /* fallthrough */
+ failed2:
```
Same remarks as for previous functions:
- send all failure cases to the failure section;
- use distinct failure labels for each failure case (even if there is no undo action between a pair of failure labels);
- name each failure label according to the action which has failed;
- only initialize local variables until you know you have succeeded.

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
+  if (ret != E_SUCCESS) goto failed2;
 
   cursor->current_positions =
       caml_stat_alloc(cursor->metadata->max_domains * sizeof(uint64_t));
```
If this allocation fails, you get an exception and all this careful failure-path code is stymied. Use `caml_stat_alloc_noexc` and then handle the failure case in the same way as all the others.

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
+ failed2:
+  caml_stat_free(runtime_events_loc);
+ failed1:
+  if (ret != E_SUCCESS) {
```
If `ret` is `E_SUCCESS` here then surely something has gone badly wrong?

NickBarnes commented on otherlibs/runtime_events/runtime_events_consumer.c:
```diff
@@ -178,70 +163,88 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
-
   *cursor_res = cursor;
+  ret = E_SUCCESS;
+  /* fallthrough */
```
It doesn't, though.

NickBarnes reviewed (changes requested):
Clearly an improvement on the previous code. A few stylistic quibbles.

gasche commented:
I believe that this PR is waiting for someone, presumably @eutro if available, to take into account the feedback of @NickBarnes. This should be good to merge afterwards.

NickBarnes commented:
In @eutro's absence, I have rebased this here: https://github.com/NickBarnes/ocaml/commit/e990bedb174a10c029b78f64c3b0ccb35b3eb2a0 and will make a further commit on that branch to reflect my review comments here.

gasche commented:
This PR was adopted by @NickBarnes in #13419 and merged. Thanks @eutro (and everyone) for the nice work!
